syntax = "proto3";

option go_package = "immortalcrab.com/sso/protos";

package dylk;

// -------------------------------------------------------------------------
// The Users service definition.
service Users {
  rpc AlterUser   (User)            returns (GeneralResponse)  {}
  rpc ListUsers   (SearchParams)    returns (UserListResponse) {}
  rpc AuthUser    (UserCredentials) returns (GeneralResponse)  {}
  rpc GetUser     (Id)              returns (UserResponse)     {}
  rpc GetCatalogs (Void)            returns (CatalogsResponse) {}
}

message User {
  int32 userId               = 1;
  string username            = 2;
  string passwd              = 3;
  int32 roleId               = 4;
  bool disabled              = 5;
  string firstName           = 6;
  string lastName            = 7;
  repeated int32 authorities = 8;
}

message CatalogEntry {
  int32 id     = 1;
  string title = 2;
}

message Authority {
  int32 id     = 1;
  int32 appId  = 2;
  string code  = 3;
  string title = 4;
}

message Param {
  string name  = 1;
  string value = 2;
}

message SearchParams {
  repeated Param paramList     = 1;
  repeated Param pageParamList = 2;
}

message UserCredentials {
  string username  = 1;
  string passwd    = 2;
}

message Id {
  int32 id = 1;
}

message Void {
}

message UserResponse {
  int32 returnCode       = 1;
  string returnMessage   = 2;
  User user              = 3;
}

message UserListResponse {
  int32 returnCode       = 1;
  string returnMessage   = 2;
  repeated User userList = 3;
  int32 totalItems       = 4;
  int32 totalPages       = 5;
}

message CatalogsResponse {
  int32 returnCode                 = 1;
  string returnMessage             = 2;
  repeated CatalogEntry roleList   = 3;
  repeated CatalogEntry appList    = 4;
  repeated Authority authorityList = 5;
}

message GeneralResponse {
  int32 returnCode     = 1;
  string returnMessage = 2;
}
